scripts/lintcommit.lua

```diff
@@ -0,0 +1,131 @@
+function commit_message_is_ok(commit_message)
+  local commit_split = vim.fn.split(commit_message, ":")
+
+  -- Return true if the type is vim-patch since most of the normal rules don't
+  -- apply.
+  if commit_split[1] == "vim-patch" then
+    return true
+  end
+
+  -- Return false if no colons are detected.
+  if vim.tbl_count(commit_split) < 2 then
+    return false
+  end
+
+  local before_colon = commit_split[1]
+  local after_colon = commit_split[2]
+
+  -- Check if commit introduces a breaking change.
+  if vim.endswith(before_colon, "!") then
+    before_colon = before_colon:sub(1, -2)
+  end
+
+  -- Check if type is correct
+  local type = vim.fn.split(before_colon, "(")[1]
+  local allowed_types = {"build", "ci", "docs", "feat", "fix", "perf", "refactor", "revert", "test", "chore"}
+  if not vim.tbl_contains(allowed_types, type) then
+    return false
+  end
+
+  -- Check if scope is empty
+  -- if before_colon:match("%(") then
+    -- scope = before_colon:match("%((.*)%)")
+  -- end
+
+  -- Check that description doesn't end with a period
+  if vim.endswith(after_colon, ".") then
+    return false
+  end
+
+  -- Check that description has exactly one whitespace after colon, followed by
+  -- a lowercase letter and then any number of letters.
+  if not string.match(after_colon, '^ %l%a*') then
+    return false
+  end
+
+  return true
+end
+
+function main()
+  local current_branch = vim.trim(vim.fn.system({'git', 'branch', '--show-current'}))
+  local ancestor_commit = vim.trim(vim.fn.system({'git', 'merge-base', 'origin/master', current_branch}))
+  local commits_str = vim.trim(vim.fn.system({'git', 'rev-list', ancestor_commit .. ".." .. current_branch}))
+
+  local commits = {}
+  for substring in commits_str:gmatch("%S+") do
+     table.insert(commits, substring)
+  end
+
+  for index, commit_hash in ipairs(commits) do
```

Comment:
Sorry to chime in so late, but any specific reason as to why not use a [commit-msg hook](https://git-scm.com/docs/githooks#_commit_msg)? It will pass the commit-message directly to `stdout` which you can pipe to `nvim` or any other script, and will even allow you to edit it, if you so wish.

There's also another hook, [pre-commit-msg](http://git-scm.com/docs/githooks#_prepare_commit_msg).

Here's the PR that I mentioned awhile ago about introducing git-hooks: https://github.com/kylo252/neovim/commit/09b12c2ab9e94c8109208cf4d672898e651ea400

We can just replace the `npx commitlint` call with this script, which means no longer depending on a local nodejs installation 😄

Reply:
> any specific reason as to why not use a [commit-msg hook](https://git-scm.com/docs/githooks#_commit_msg)?

how will that help when running in contexts like CI ? It is just less flexible.

off topic for this PR.